include/fastdds/rtps/builtin/discovery/endpoint/EDP.h

```diff
@@ -350,14 +356,14 @@ class EDP
     struct reader_status
     {
         fastdds::dds::SubscriptionMatchedStatus subscription_matched;
-        fastdds::dds::QosPolicyId_t requested_incompatible_qos;
+        fastdds::dds::PolicyMask requested_incompatible_qos;
     };
     std::map<GUID_t, reader_status> reader_status_;
 
     struct writer_status
     {
         fastdds::dds::PublicationMatchedStatus publication_matched;
-        fastdds::dds::QosPolicyId_t offered_incompatible_qos;
+        fastdds::dds::PolicyMask offered_incompatible_qos;
```

Comment:
Looking at this again, I think it is a better solution to keep the data structures as they were, and extend the `validMatching` methods.

I mean, create two new methods having the policy mask as an output parameter, and making the old versions call the new one with the mask as a local variable

```
bool valid_matching(
        const WriterProxyData* wdata,
        const ReaderProxyData* rdata,
        PolicyMask& offered_incompatible_qos);

bool valid_matching(
        const ReaderProxyData* rdata,
        const WriterProxyData* wdata,
        PolicyMask& requested_incompatible_qos);
```

This way we save memory, as well as improve multi-threading behavior.

Reply:
As per our conversation, I also added a mask to return the cause of failure to match.

Just one thing more, the mask could potentially return several causes for not matching. But I don't think all of them are even interesting. For example, having a different Topic name already disqualifies for all the rest. Do we want to support returning several active bits on the mask? If so, which ones?

Aborting the matching algorithm as soon as one cause is found may save some time, but I guess the only 'normal' reason is to have different Topic names, and all the rest could be considered 'probably a configuration error', and could be aggregated. What do you think?